gufe/components/proteincomponent.py

```diff
@@ -1,86 +1,671 @@
+
+        # Atoms
+        atoms = {}
+        for atom in sorted(dict_prot["atoms"], key=lambda x: x[5]["id"]):
+            aid = atom[5]["id"]
+            atom_mi_dict = atom[6]
+            chainn = atom_mi_dict["ChainId"]
+            resInd = atom_mi_dict["ResidueNumber"]
+
+            key = str(chainn) + "_" + str(resInd)
+            r = residues[key]
+
+            atom = top.addAtom(
+                name=atom[1],
+                residue=r,
+                id=aid,
+                element=app.Element.getByAtomicNumber(atom[0]),
+            )
+            atoms[atom.index] = atom
+
+        # Bonds
+        for bond in dict_prot["bonds"]:
+            top.addBond(
+                atom1=atoms[bond[0]],
+                atom2=atoms[bond[1]],
+                type=bond[2],
+                order=bond[2],
+            )
+
+        # Geometrics
+        if dict_prot["molecules"]["unit_cell_dimensions"] != "None":
+            top.setUnitCellDimensions(
+                np.array(dict_prot["molecules"]["unit_cell_dimensions"])
+                * omm_unit.angstrom
+            )
+        else:
+            top.setUnitCellDimensions(None)
+
+        if dict_prot["molecules"]["periodic_box_vectors"] != "None":
+            top.setPeriodicBoxVectors(
+                list(
+                    map(
+                        lambda x: np.array(x),
+                        dict_prot["molecules"]["periodic_box_vectors"],
+                    )
+                )
+                * omm_unit.angstrom
+            )
+        else:
+            top.setPeriodicBoxVectors(None)
+
+        return top
+
+    def to_openmm_positions(self) -> omm_unit.Quantity:
+        """
+        serialize the positions to openmm.unit.Quantity
+        ! only one frame at the moment!
+
+        Returns
+        -------
+        omm_unit.Quantity
+            Quantity containing protein atom positions
+        """
+        np_pos = deserialize_numpy(self.to_dict()["conformers"][0])
+        openmm_pos = (
+            list(map(lambda x: np.array(x), np_pos)) * omm_unit.angstrom
+        )
+
+        return openmm_pos
+
+    def to_pdbFile(self, out_path: Union[str, io.TextIOWrapper] = None) -> str:
+        """
+        serialize protein to pdb file.
+
+        Parameters
+        ----------
+        out_path : str
+            provide path or FileIO to the resulting file, by default None
+
+        Returns
+        -------
+        str
+            string path to the resulting pdb.
+        """
+        # get top:
+        openmm_top = self.to_openmm_topology()
+
+        # get pos:
+        openmm_pos = self.to_openmm_positions()
+
+        # write file
+        if isinstance(out_path, str):
+            out_file = open(out_path, "w")
+        elif isinstance(out_path, io.TextIOWrapper):
+            out_file = out_path
+            out_path = str(out_file.name)
+        else:
+            raise ValueError("Out path type was not as expected!")
```

Comment:
This breaks the ability to use `io.StringIO`.

I think you really just want something here like:

```python
if not isinstance(out_path, io.TextIOBase):
    # allows pathlike/str; we close on completion
    out_path = open(out_path, mode='w')
    must_close = True

PDBFile.writeFile(open_top, openmm_pos, file=out_path)

if must_close:
    # we only close the file if we had to open it
    out_path.close()
```

Reply:
hm on TextIOBase I get a mypy error:
![image](https://user-images.githubusercontent.com/12428005/191014220-e93b3218-b377-43d2-a22d-be1cfdeee3ee.png)

However I'm not sure what the TextIO type should be, as it seems not to be accesible???